angr/state_plugins/heap/heap_base.py

```diff
@@ -0,0 +1,118 @@
+        problems that any heap implementation will face: how to decide the amount of space to allocate upon request for
+        a symbolic size. Either we do as we do here and silently constrain the amount returned to a default max value,
+        or we could add a path constraint to the state to prevent exploration of any paths that would have legitimately
+        occurred given a larger allocation size.
+
+        The first approach (the silent maximum) has its benefit in that the explored state space will not be
+        constrained. Sometimes this could work out, as when an allocation is returned that is smaller than requested but
+        which the program doesn't end up making full use of anyways. Alternatively, this lack of fidelity could cause
+        the program to overwrite other allocations made, since it should be able to assume the allocation is as large as
+        it requested it be.
+
+        The second approach (the path constraint) has its benefit in that no paths will be explored that *could* fail
+        when an allocation is made too small. On the other hand, as stated above, some of these paths might not have
+        failed anyways, and doing this causes us to lose the opportunity to explore those paths.
+
+        Perhaps these behaviors could be parameterized in the future?
+        """
+        if self.state.solver.symbolic(sim_size):
+            size = self.state.solver.max_int(sim_size)
+            if size > self.state.libc.max_variable_size:
+                l.warning("Allocation request of %d bytes exceeded maximum of %d bytes; allocating %d bytes",
+                          size, self.state.libc.max_variable_size, size)
+                size = self.state.libc.max_variable_size
+        else:
+            size = self.state.solver.eval(sim_size)
+        return size
+
+    def _malloc(self, sim_size):
+        """
+        Handler for any libc `malloc` SimProcedure call. If the heap has faithful support for `malloc`, it ought to be
+        implemented in a `malloc` function (as opposed to the `_malloc` function).
+
+        :param sim_size: the amount of memory (in bytes) to be allocated
+        """
+        raise NotImplementedError("%s not implemented for %s" % (self._malloc.__func__.__name__,
+                                                                 self.__class__.__name__))
+
+    def _free(self, ptr):
+        """
+        Handler for any libc `free` SimProcedure call. If the heap has faithful support for `free`, it ought to be
+        implemented in a `free` function (as opposed to the `_free` function).
+
+        :param ptr: the location in memory to be freed
+        """
+        raise NotImplementedError("%s not implemented for %s" % (self._free.__func__.__name__,
+                                                                 self.__class__.__name__))
+
+    def _calloc(self, sim_nmemb, sim_size):
```

Comment:
Why is there an explicit calloc? Is there any benefit to having calloc done internally instead of pushing it out to the simprocedure which calls malloc?

Reply:
Good question. At the time I only did it to be consistent with the treatment the other functions were getting.

If I'm understanding correctly, you're asking whether it would be possible to implement `calloc` in a common manner for the heaps we have right now (by implementing it in terms of `malloc`). I'm pretty sure that would be fine right now.

Reply:
We might want to check that `mmap_base` is not overlapping with the heap:
`self.mmap_base = DEFAULT_HEAP_LOCATION + DEFAULT_HEAP_SIZE * 2` 
Since heap_location, was set to the default location, this was implicitly assumed so far.

Reply:
Good catch. Probably deserves better treatment than [this](https://github.com/angr/angr/pull/1432/commits/c8a856677f2dc9fe83ae2f9f24dafe629f1515a5) in the long run, but it should be a bit more sane with respect to the possibility you bring up.